R/kwic.R

```diff
@@ -83,44 +82,45 @@ kwic.corpus <- function(x, pattern, window = 5,
 #'          "Is it a question?",
 #'          "Sometimes you don't know if this is it.",
 #'          "Is it a bird or a plane or is it a train?")
-#' kwic(txt, c("is", "a"), valuetype = "fixed")
+#' kwic(txt, "is")
+#' kwic(txt, "in", valuetype = "regex")
 #' kwic(txt, phrase(c("is", "a", "is it")), valuetype = "fixed")
 #'
 #' toks <- tokens(txt)
-#' kwic(toks, c("is", "a"), valuetype = "fixed")
-#' kwic(toks, phrase(c("is", "a", "is it")), valuetype = "fixed")
+#' kwic(txt, "is", valuetype = "regex")
+#' kwic(txt, "or", window = 2)
 #'
 #' corp <- corpus(txt)
-#' kwic(corp, c("is", "a"), valuetype = "fixed", separator = "", remove_separators = FALSE)
+#' print(kwic(corp, "is"), separator = "")
 #' @export
 kwic.tokens <- function(x, pattern, window = 5,
```

Comment:
Shall we deprecate window and separator to make clear that `kwic()` is only for locating patters?

Reply:
I would say no, simply because this is the way that users have gotten used to using kwic since it was first written. It's also more natural to put this into the function that creates the kwic. We can make it clear that this can be changed using `as.data.frame()` or `print()`, but most people using quanteda probably don't want to type `print()` to see output, they just want to see default output. My experience with `kwic()` in particular is that this is a function most used by our most qualitative and interactive users.

When I use `kwic()` or teach this, I also prefer to set it at the time we call it. And this is not like what we've discussed about removing the convenience options from `dfm()` or removing `dfm.character()`, since those are a form of bad practice, skipping a step that the user should be taking more direct responsibility for. Rather, the `window` and `separator` options belong to `kwic()` and are natural to associate with creating a `kwic()`, but our implementation has now disassociated itself from needing to set this at creation time. But that's more about our current implementation than what the action implemented by the function calls for.

One more reason: If we do change `kwic()` again in the future, it might make more sense to set these at creation time. 

So I think it makes sense to keep them without warnings. Most users will happily never know the difference.

Reply:
It makes more sense to have `window` in `kwic()` but I am not comfortable with the same argument in the different places. How about let users to call `kwic.kwck()` to change `window` or `separator`? `max_nrow` should still be in `print()`.

Reply:
`kwic.kwic()` makes no sense unless it would be possible to search a `kwic()` for keywords - although we could make it do so since we have the tokens stored inside it. I think that's an edge case however that will create more confusion than the benefit that might come from using such a method to (re)set window or separator.

I think it's ok to keep window and separator in both the constructor and the print and conversion methods. We just have to make it clear that the default is set when the kwic is created, but can be changed at the print or data.frame stage. My guess is that it will be a very rare user that wants to do so, so in the vast majority of cases for the vast majority of users, the kwic functions will simply function as they have before.

For `print.kwic()` I'd suggest that our default be `max_nrow = -1`. Want me to implement that?

Reply:
Let's take out `window` or `separator` from print for now. My understanding is that people do not want to run `kwic()` multiple times because it was really slow for large object, but it is no loner so was far as the resulting object is subset. Let set `max_nrow = 1000` or a bit more so that `kwic()` does not freeze.

Reply:
Thinking about why we have to make a decision to add these argument to either or both `kwic()` or `print()`... I think my design to make `print.kwic()` was the mistake. `kwic()` should be the print function that creates the concordance view.

What I should have done is creating a function that might be call `search()` or `locate()` that returns the positions of matches (the same as what `kwic.tokens()` does in this branch) and calling this function in `kwic()`. If we add a special class to this object (like "position"), we can make a `kwic.position()`. The object can be useful for information retrieval applications too.

Reply:
That would work too. But the current structure is fine. Let's do this:
- Keep `kwic()` as in this PR.
- Remove `window` and `separator` from `print.kwic()`, but keep these for `as.data.frame.kwic()`.
- Add `max_nrow = 1000` to `print.kwic()` and make it work like the other object print functions (with the "x more rows..." output).
- Not implement yet the other functions you mention above, but keep in mind that these can be added when and if needed.

Reply:
OK. That is the middle way.

Reply:
Alright, I did all of the above in 13fa3ea. This implements the same print options now for kwic as for other core objects, including `show_summary`. The arguments `window` and `separator` are kept in `as.data.frame.kwic()` but removed from  `print.kwic()` (but I left the internal code for these intact in case we want to add them back).

We can remove the .corpus and .character methods in a separate PR, when we address #2058 more directly.

Reply:
Alright, I did all of the above in 13fa3ea. This implements the same print options now for kwic as for other core objects, including `show_summary`. The arguments `window` and `separator` are kept in `as.data.frame.kwic()` but removed from  `print.kwic()` (but I left the internal code for these intact in case we want to add them back).

We can remove the .corpus and .character methods in a separate PR, when we address #2058.